src/lars.cpp

```diff
@@ -63,5 +64,5 @@ SEXP UnserializeLARSPtr(Rcpp::RawVector str)
     boost::archive::binary_iarchive ia(iss);
     ia >> boost::serialization::make_nvp("LARS", *ptr);
   }
-  return Rcpp::wrap((XPtrLARS) ptr);
+  return std::move((XPtrLARS) ptr);
```

Comment:
Love this. 

If this works, maybe also change the function return from `SEXP` to the `XPtrLARS` typedef or templated `XPtr`?

Reply:
But Rcpp::XPtr<LARS> will not be recognised in RcppExports.cpp, because we are not including `#include <mlpack/methods/lars/lars_main.cpp>` there.

Reply:
Ah, yes.  Good point. 

You can (if you want) arrange for that to work.  One well-defined (but little-known) trick it to have a header file in `inst/include` with the name of the R package plus  `_types.h` appended to it.  One example is [this file for RcppQuantuccia](https://github.com/eddelbuettel/rcppquantuccia/blob/master/inst/include/RcppQuantuccia_types.h).

Such a file will be convention be included in the (auto-generated) `RcppExports.cpp`.

It's a minor detail but it makes the code more C++-ish, I find.  OTOH the C interface from R wants a `SEXP`.  "Both forms are correct."

Reply:
I can do this if you want. No problem :).

Reply:
Your code, your style.  As I wrote a case can be made for either.  Because `SEXP` is a union-type it is sometimes less informative.  Here the body is short so it is obvious.

Reply:
Hi @eddelbuettel, I have checked, If all we include all the `main.cpp` in `RcppMLPACK_types.h` it will be create redefinition error, As all the `main.cpp` contain a `mlpackMain()` function. If you agree we can add a comment here regarding the return type.

Thank You.

Reply:
I am sorry but where did I ask you to copy main.cpp into rcppmlpack_types.h?  

I even pointed you to a working file, and obviously such a type-defining file will be included multiple times [hey that is *why* we do thos to tell RcppExports] so you can't have actual definitions.

Reply:
I am really really sorry, I was doing some blunder here. I will definitely work on this part later point of time.

Reply:
No worries -- it's tricky. I am just looking at the _exact same issue_ in RcppGSL (ticket [here](https://github.com/eddelbuettel/rcppgsl/issues/25)) which _somehow_ never raised its head for us using `g++` or `clang++` but appears to now bite on Solaris.  And it is clearly buggy from my side: actual definitions in the header, not using `inline`. Bam. Linker error on multiple symbols.

That said _for simple declarations_ like the `typedef` here the mechanism of `packagename_types.h` is great, and simple.  But as always in C++ it is also easy to go overboard and shoot one's foot (and, as famously said, leg) off with it.  

But when it works it of course just works :)

Reply:
@Yashwants19 one idea here would be to do something similar to the changes you just made for the Go bindings, where all of the model types are "collected" by CMake and put into their own file. :+1:  Anyway, I have no opinion for what the right way to go here is (I agree with @eddelbuettel that it's up to you), but hopefully the comment is helpful. :)

Reply:
Yeah, This sound good. I will use CMake for collecting all the model types and put them into their own file.